packages/next-auth/src/lib/index.ts

```diff
@@ -0,0 +1,239 @@
+     *
+     * You can override this behavior by returning a {@link NextResponse}.
+     *
+     * @example
+     * ```ts title="app/auth.ts"
+     * ...
+     * async authorized({ request, auth }) {
+     *   const url = request.nextUrl
+     *
+     *   if(request.method === "POST") {
+     *     const { authToken } = (await request.json()) ?? {}
+     *     // If the request has a valid auth token, it is authorized
+     *     const valid = await validateAuthToken(authToken)
+     *     if(valid) return true
+     *     return NextResponse.json("Invalid auth token", { status: 401 })
+     *   }
+     *
+     *   // Logged in users are authenticated, otherwise redirect to login page
+     *   return !!auth.user
+     * }
+     * ...
+     * ```
+     *
+     * :::warning
+     * If you are returning a redirect response, make sure that the page you are redirecting to is not protected by this callback,
+     * otherwise you could end up in an infinite redirect loop.
+     * :::
+     */
+    authorized?: (params: {
+      /** The request to be authorized. */
+      request: NextRequest
+      /** The authenticated user or token, if any. */
+      auth: Session | null
+    }) => Awaitable<boolean | NextResponse | undefined>
+  }
+}
+
+function detectOrigin(_headers: Headers | ReturnType<typeof headers>) {
+  const host = _headers.get("x-forwarded-host") ?? _headers.get("host")
+  const protocol =
+    _headers.get("x-forwarded-proto") === "http" ? "http" : "https"
+  return new URL(process.env.NEXTAUTH_URL ?? `${protocol}://${host}`)
+}
+
+/** Server-side method to read the session. */
+async function getSession(headers: Headers, config: NextAuthConfig) {
+  const origin = detectOrigin(headers)
+  const request = new Request(`${origin}/api/auth/session`, {
+    headers: { cookie: headers.get("cookie") ?? "" },
+  })
+  config.useSecureCookies ??= origin.protocol === "https:"
+
+  // Since we are server-side, we don't need to filter out the session data
+  // See https://nextjs.authjs.dev/v5#authenticating-server-side
+  config.callbacks = Object.assign(config.callbacks ?? {}, {
+    session({ session, user, token }) {
+      return { ...session, user: user ?? token }
+    },
+  })
+  return Auth(request, config)
+}
+
+export interface NextAuthRequest extends NextRequest {
+  auth: Session
+}
+
+export type NextAuthMiddleware = (
+  request: NextAuthRequest,
+  event: NextFetchEvent
+) => ReturnType<NextMiddleware>
+
+export type WithAuthArgs =
+  | [NextAuthRequest, any]
+  | [NextAuthMiddleware]
+  | [AppRouteHandlerFn]
+  | [NextApiRequest, NextApiResponse]
+  | [GetServerSidePropsContext]
+  | []
+
+function isReqWrapper(arg: any): arg is NextAuthMiddleware | AppRouteHandlerFn {
+  return typeof arg === "function"
+}
+
+export function initAuth(config: NextAuthConfig) {
+  return (...args: WithAuthArgs) => {
+    if (!args.length) return getSession(headers(), config).then((r) => r.json())
+    if (args[0] instanceof Request) {
```

Comment:
First, thank you for all your open source work! I've been using next-auth since v3 and it's great.
I'm trying to use this but I run into this error what if `args[0] instanceof Request` is true but instead of being `middleware.ts` is an edge function? 
How could I get the session using `auth` in an edge function where I only have the request and not the response?
I didn't find an overload where it only uses `Request` and if i try to use it without any arguments (`auth()`) I get this error `Invariant: Method expects to have requestAsyncStorage, none available` which seems to be related to this https://github.com/vercel/next.js/issues/46356

Reply:
You are expected to wrap the handler, see: https://auth-docs-git-feat-nextjs-auth-authjs.vercel.app/guides/upgrade-to-v5#authenticating-server-side

We need this so we control the outgoing response headers (so we can rotate the session expiry)

Reply:
Thank you! I'm using this with trpc and it worked perfectly, really looking forward to this PR so we can start using next-auth in the edge.